src/lib/scheduler/worker.cpp

```diff
@@ -128,17 +138,54 @@ void Worker::join() {
 uint64_t Worker::num_finished_tasks() const { return _num_finished_tasks; }
 
 void Worker::_wait_for_tasks(const std::vector<std::shared_ptr<AbstractTask>>& tasks) {
-  auto tasks_completed = [&tasks]() {
-    // Reversely iterate through the list of tasks, because unfinished tasks are likely at the end of the list.
-    for (auto it = tasks.rbegin(); it != tasks.rend(); ++it) {
-      if (!(*it)->is_done()) {
+  // This lambda checks if all tasks from the vector have been executed. If they are, it causes _wait_for_tasks to
+  // return. If there are remaining tasks, it primarily tries to execute these. If they cannot be executed, the
+  // worker performs work for others (i.e., executes tasks from the queue).
+  auto check_own_tasks = [&]() {
+    auto all_done = true;
+    for (auto it = tasks.begin(); it != tasks.end(); ++it) {
+      const auto& task = *it;
+      if (task->is_done()) {
+        continue;
+      }
+
+      // Task is not yet done - check if it is ready for execution
+      if (!task->is_ready()) {
+        all_done = false;
+        continue;
+      }
+
+      // Give other tasks a certain chance of being executed, too. Anectotal evidence says that this is a good idea.
+      // For some reason, this keeps the memory consumption of TPC-H Q6 low even if the scheduler is overcommitted.
```

Comment:
This randomly doing other work is the actual trick here to get rid of the waiting chains, right?
How does this comment on Q6 relate to that?

Reply:
No, the "waiting chains" are fixed by the cases in which we do our own work. However, if we *only* do our work, the memory consumption of Q6 explodes. I have no idea why.

Any idea how to improve this comment? I worry that going to much into detail here would confuse more than help, given that future readers do not necessarily know the history of this change.

Reply:
Could you extend "other tasks" a little bit more? The reader might wonder what you don't use `continue`.

Reply:
"Give other tasks, i.e., tasks that are unrelated to what we are currently waiting on, a certain chance of being executed, too"

Reply:
Jeop.